### Data Analysis with Panda
- Bike Sales
- Michael Hendricks
- July 20

In [59]:
import pandas as pd
import numpy as np

In [60]:
station_df = pd.read_csv('data/station.csv')
trip_df = pd.read_csv('data/trip.csv', error_bad_lines=False)
weather_df = pd.read_csv('data/weather.csv')

b'Skipping line 50794: expected 12 fields, saw 20\n'


In [61]:
trip_df[['tripduration']].mean()

tripduration    1178.295675
dtype: float64

In [62]:
avg_trip_duration = trip_df[['tripduration']].mean()[0]

In [63]:
avg_trip_duration

1178.2956753399776

### 1. The average trip duration for a borrowed bicycle is 1178.295675

In [64]:
trip_df['birthyear'].mode()

0    1987.0
dtype: float64

In [65]:
most_common_renter_birthyear = trip_df['birthyear'].mode()[0]

In [66]:
most_common_renter_birthyear

1987.0

### 2. The most common birth year of a bicycle sharer is 1987, and since that's the year I was born, that makes them 33

In [67]:
#weather_df['Date'] = pd.to_datetime(weather_df['Date'], format='%m/%d/%Y')
#weather_df.loc[weather_df['Date'].dt.month == 12][['Date', 'Precipitation_In']].mean()

In [68]:
prec_per_day = weather_df[['Date','Precipitation_In']].mean()

In [69]:
prec_per_month = prec_per_day * 30

In [70]:
avg_monthly_prec = prec_per_month[0]

In [71]:
avg_monthly_prec

3.1519593613933234

### 3. The average precipitation per month is 3.1519593613933234

In [72]:
#station_df[['name','current_dockcount']]
station_df[['current_dockcount']].mean()

current_dockcount    16.517241
dtype: float64

In [73]:
avg_bikes_at_station = station_df[['current_dockcount']].mean()[0]

In [74]:
avg_bikes_at_station

16.517241379310345

### 4. The average number of bikes at a given station is 16.517241

In [75]:
# notnull() filters out NaN returns
modified_stations = station_df[station_df['modification_date'].notnull()][['name','install_dockcount', 'modification_date', 'current_dockcount']].set_index('name')

In [76]:
pre_install = modified_stations[['install_dockcount']].mean()

In [77]:
pre_install[0]

17.529411764705884

In [78]:
post_install = modified_stations[['current_dockcount']].mean()

In [79]:
post_install[0]

13.764705882352942

### 5. When a bike station is modified it is more likely that it's lose bikes because the average dock count of stations before modifications was higher than the average dock count of those same stations after modifications

In [80]:
weather_df[['Max_Temperature_F']].max()

Max_Temperature_F    98
dtype: int64

### 6. The hottest day on record hit a max of 98 degrees F

In [81]:
weather_df[['Min_TemperatureF']].min()

Min_TemperatureF    23
dtype: int64

### 7. The coldest day on record got as cold as 23 defrees F

In [82]:
weather_df[['Mean_Visibility_Miles']].mean()

Mean_Visibility_Miles    9.429608
dtype: float64

### 8. The overall avg visibility is 9.42 miles

In [85]:
def test():

    def assert_equal(actual,expected):
        assert actual == expected, f"Expected {expected} but got {actual}"

    assert_equal(trip_df[['tripduration']].mean()[0], 1178.2956753399776)
    assert_equal(trip_df['birthyear'].mode()[0], 1987.0)
    assert_equal(avg_monthly_prec, 3.1519593613933234)
    assert_equal(station_df[['current_dockcount']].mean()[0], 16.517241379310345)
    assert_equal(pre_install[0] > [post_install[0]], True)
  

    print("Success!!")

In [86]:
test()

Success!!
